language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Exception.hs

```diff
@@ -30,17 +69,72 @@ data MachineError err
     | ConstAppMachineError ConstAppError
       -- ^ An attempt to compute a constant application resulted in 'ConstAppError'.
     | OtherMachineError err
-    deriving (Eq)
+    deriving (Show, Eq)
+
+data EvaluationError internal user
+    = InternalEvaluationError (MachineError internal)
+      -- ^ Indicates bugs.
+    | UserEvaluationError user
+      -- ^ Indicates user errors.
+    deriving (Show, Eq)
+
+mtraverse makeClassyPrisms
+    [ ''UnliftingError
+    , ''ConstAppError
+    , ''MachineError
+    , ''EvaluationError
+    ]
+
+instance AsMachineError (EvaluationError internal user) internal where
+    _MachineError = _InternalEvaluationError
+instance AsConstAppError (MachineError err) where
+    _ConstAppError = _ConstAppMachineError
+instance AsConstAppError (EvaluationError internal user) where
+    _ConstAppError = _InternalEvaluationError . _ConstAppMachineError
+instance AsUnliftingError ConstAppError where
+    _UnliftingError = _UnliftingConstAppError
+instance AsUnliftingError (EvaluationError internal user) where
+    _UnliftingError = _InternalEvaluationError . _UnliftingConstAppError
+instance AsUnliftingError (MachineError err) where
+    _UnliftingError = _ConstAppMachineError . _UnliftingConstAppError
+
+data ErrorWithCause err
+    = ErrorWithCause err (Maybe (Term TyName Name ()))
+    deriving (Eq, Functor)
+
+type MachineException internal = ErrorWithCause (MachineError internal)
+type EvaluationException internal user = ErrorWithCause (EvaluationError internal user)
 
--- | The type of exceptions an abstract machine can throw.
-data MachineException err = MachineException
-    { _machineExceptionError :: MachineError err     -- ^ An error.
-    , _machineExceptionCause :: Term TyName Name ()  -- ^ A 'Term' that caused the error.
-    } deriving (Eq)
+-- | "Prismatically" throw an error and its (optional) cause.
+throwingWithCause
+    :: MonadError (ErrorWithCause e) m
+    => AReview e t -> t -> Maybe (Term TyName Name ()) -> m x
+throwingWithCause l t cause = reviews l (\e -> throwError $ ErrorWithCause e cause) t
```

Comment:
I'm way too lazy to look into the source of `lens` and copypaste something from there, unless there is a reason to do so:

```haskell
throwing _UnliftingError (ErrorWithCause "Not a builtin ()" $ Just term)
```

gives

```
    • Couldn't match expected type ‘UnliftingError’
                  with actual type ‘ErrorWithCause [Char]’
    • In the second argument of ‘throwing’, namely
        ‘(ErrorWithCause "Not a builtin ()" $ Just term)’
      In the expression:
        throwing
          _UnliftingError (ErrorWithCause "Not a builtin ()" $ Just term)
```

I don't know if it's possible to compose a prism and a lens and get a review to fit this into what `throwing` expects, but it seems to me it's not possible and it would be a harder thing anyway. So I think what is there is pretty simple.

Reply:
Oh I see, the cause goes on the outside of the final error, rather than inside the error that gets injected. That is indeed a bit fiddly.